module/os/linux/spl/spl-taskq.c

```diff
@@ -1262,6 +1263,105 @@ taskq_destroy(taskq_t *tq)
 }
 EXPORT_SYMBOL(taskq_destroy);
 
+static void taskq_sync_assign(void *arg);
+
+typedef struct taskq_sync_arg {
+	kthread_t	*tqa_thread;
+	kcondvar_t	tqa_cv;
+	kmutex_t 	tqa_lock;
+	int		tqa_ready;
+} taskq_sync_arg_t;
+
+static void
+taskq_sync_assign(void *arg)
+{
+	taskq_sync_arg_t *tqa = arg;
+
+	mutex_enter(&tqa->tqa_lock);
+	tqa->tqa_thread = curthread;
+	tqa->tqa_ready = 1;
+	cv_signal(&tqa->tqa_cv);
+	while (tqa->tqa_ready == 1)
+		cv_wait(&tqa->tqa_cv, &tqa->tqa_lock);
+	mutex_exit(&tqa->tqa_lock);
+}
+
+/*
+ * Create a taskq with a specified number of pool threads. Allocate
+ * and return an array of kthread_t pointers, one for each thread
+ * in the pool. Note, the array is not ordered.
+ *
+ * A taskq created with this function may only be destroyed by
+ * taskq_destroy_synced().
+ */
+taskq_t *
+taskq_create_synced(const char *name, int nthreads, pri_t pri,
+    int minalloc, int maxalloc, uint_t flags, kthread_t ***ktpp)
+{
+	taskq_t *tq;
+	taskq_sync_arg_t *tqs = kmem_zalloc(sizeof (*tqs) * nthreads, KM_SLEEP);
+	kthread_t **kthreads = kmem_zalloc(sizeof (*kthreads) * nthreads,
+	    KM_SLEEP);
+
+	flags &= ~(TASKQ_DYNAMIC | TASKQ_THREADS_CPU_PCT | TASKQ_DC_BATCH);
+
+	tq = taskq_create(name, nthreads, minclsyspri, nthreads, INT_MAX,
+	    flags | TASKQ_PREPOPULATE);
+	VERIFY(tq != NULL);
+	VERIFY(tq->tq_nthreads == nthreads);
+
+	/* spawn all syncthreads */
+	for (int i = 0; i < nthreads; i++) {
+		cv_init(&tqs[i].tqa_cv, NULL, CV_DEFAULT, NULL);
+		mutex_init(&tqs[i].tqa_lock, NULL, MUTEX_DEFAULT, NULL);
+		(void) taskq_dispatch(tq, taskq_sync_assign,
+		    &tqs[i], TQ_FRONT);
+	}
+
```

Comment:
At least for the `module/os/linux/spl/spl-taskq.c` and `lib/libzpool/taskq.c`  taskq implementations `taskq_create()` will spawn all of the threads before exiting.  You should be able to replace lines 1321-1345 with this one liner, and then remove all the support code `taskq_sync_assign()` and `struct taskq_sync_arg`.

```c
*ktpp = tq->tq_threadlist;
```

Reply:
On Linux I don't see the taskq_t has a tq_threadlist member.  It does have tq_thread_list but that is a struct list_head. Is there another way to get the curthread for each pool thread (other than from the thread itself running)?

Reply:
The `taskq->tq_thread_list` is a list of `taskq_thread_t` structs linked together with the embedded `tqt->tqt_thread_list` list_head member.  To get all of the `kthread_t *`'s you can walk the list and dump each `tqt->tqt_thread` member.  On Linux a `kthread_t` is defined to be a `struct task_struct`.  When `curthread` is called in a taskq function one of these `kthread_t` / `struct task_struct`'s will be returned.

Sorry, it's not quite a one liner as I mistakenly said above.